### ISSUE TITLE
How can I train a qwen eagle model?

### ISSUE THREAD

**Issue by Ximingwang-09** [NickL77/BaldEagle#4]
I can't find qwen in modules.model. It's only support llama_eagle.

**Comment by KerwinKai** [comment]
For generate data, i think we only need to change llama ckpt path to qwen ckpt path in https://github.com/NickL77/BaldEagle/blob/master/generate_data/generate_data.py#L76

**Comment by KerwinKai** [comment]
> > For generate data, i think we only need to change llama ckpt path to qwen ckpt path in https://github.com/NickL77/BaldEagle/blob/master/generate_data/generate_data.py#L76
> 
> I think we also need a QwenForCausalLMEagle like 'LlamaForCausalLMEagle' ?

QwenForCausalLMEagle is needed during inference. For training, we only need to copy the layer of the base model to train. I took a simpler approach during inference, that is, using LlamaForCausalLMEagle to represent QwenForCausalLMEagle. I only need to ensure that the config of the draft model is the same.

**Comment by Ximingwang-09** [comment]
> > > For generate data, i think we only need to change llama ckpt path to qwen ckpt path in https://github.com/NickL77/BaldEagle/blob/master/generate_data/generate_data.py#L76
> > 
> > 
> > I think we also need a QwenForCausalLMEagle like 'LlamaForCausalLMEagle' ?
> 
> QwenForCausalLMEagle is needed during inference. For training, we only need to copy the layer of the base model to train. I took a simpler approach during inference, that is, using LlamaForCausalLMEagle to represent QwenForCausalLMEagle. I only need to ensure that the config of the draft model is the same.

Thanks, I'll try it again!

**Comment by Arcmoon-Hu** [comment]
> For generate data, i think we only need to change llama ckpt path to qwen ckpt path in https://github.com/NickL77/BaldEagle/blob/master/generate_data/generate_data.py#L76

And need change regex part? Because the qwen's special token is different from llama

**Comment by NickL77** [comment]
You're right, you'd also need to change this to match chat template: https://github.com/NickL77/BaldEagle/blob/master/generate_data/generate_data.py#L79-L80

**Comment by Greatpanc** [comment]
When I replace the model with Qwen2.5-7B, the train logging_valid_token_count 0 and top_1_acc_sum 0, where am I using it wrong? ![Image](https://github.com/user-attachments/assets/dca09c76-3ad7-41ce-be84-af6921a4cf60)

**Comment by NickL77** [comment]
Did you change the chat template headers to Qwen's instead of Llama's? https://github.com/NickL77/BaldEagle/blob/master/generate_data/generate_data.py#L79-L80

It seems like everything is masked. You can see if the masking is correct using the `view_data.py` script: https://github.com/NickL77/BaldEagle/blob/master/generate_data/README.md

**Comment by Greatpanc** [comment]
![Image](https://github.com/user-attachments/assets/454b64e9-0607-4883-879d-52d8028078aa) This is the result I ran through the view_data.py, I am not very good at seeing whether this interface is correct

**Comment by Greatpanc** [comment]
Thanks, I'll just change it like this

What do top_1_acc, top_2_acc and top_3_acc mean?

**Comment by Greatpanc** [comment]
> Is the target token within the top_1, top_2, or top_3 highest probabilities of the prediction. It is similar to topk parameter when running inference

This value will have an acceleration effect. I have only trained it to about 0.1 so far. I don't know if it's due to insufficient data

**Comment by Greatpanc** [comment]
> I'm not sure why you're getting OOM. I can train on a 24GB 3090. Are you training a larger draft model? Or using larger per_device_batch_size?

No，this is my train code，But when I located it, I found that the batchsize in the forward would be related to my card number

> import json
> import os
> import torch
> import wandb
> import random
> 
> from safetensors import safe_open
> 
> from transformers.models.llama.configuration_llama import LlamaConfig
> 
> from transformers import AutoTokenizer, TrainingArguments
> 
> from modules.model.llama_eagle import LlamaForCausalLMEagle
> from modules.data.data import (
>     EagleLocalDataset,
>     DataCollatorWithPadding,
>     AddUniformNoise,
>     list_local_files,
> )
> from modules.trainer.trainer import EagleTrainer
> 
> wandb.init(project="BaldEagle")
> wandb_run_name = wandb.run.name
> 
> path = "models_zoo/Qwen/Qwen2.5-7B-Instruct"
> 
> # -------------------------------- Load original Llama weights --------------------------------
> 
> with open(os.path.join(path, "model.safetensors.index.json"), "r") as f:
>     index_json = json.loads(f.read())
>     emb_path = index_json["weight_map"]["model.embed_tokens.weight"]
>     lm_head_path = index_json["weight_map"]["lm_head.weight"]
> 
> with safe_open(os.path.join(path, emb_path), framework="pt", device="cpu") as f:
>     tensor_slice = f.get_slice("model.embed_tokens.weight")
>     vocab_size, hidden_dim = tensor_slice.get_shape()
>     tensor = tensor_slice[:, :hidden_dim]
> 
> with safe_open(os.path.join(path, lm_head_path), framework="pt", device="cpu") as f:
>     lm_head_weights = f.get_slice("lm_head.weight")[:, :]
> 
> 
> # -------------------------------- Create draft model + tokenizer + head --------------------------------
> 
> tokenizer = AutoTokenizer.from_pretrained(path)
> tokenizer.pad_token = tokenizer.eos_token
> 
> print("vocab_size: ", vocab_size)
> print("hidden_dim: ", hidden_dim)
> model_args = LlamaConfig(
>     vocab_size=vocab_size,
>     hidden_size=hidden_dim,
>     intermediate_size=14336,
>     num_hidden_layers=1,
>     bos_token_id=128000,
>     eos_token_id=[128001, 128008, 128009],
>     num_key_value_heads=7,
>     num_attention_heads=28,
>     tie_word_embeddings=False,
> )
> 
> draft_model = LlamaForCausalLMEagle(model_args)
> draft_model.load_embedding_weights(tensor)
> draft_model.to("cuda:0")
> draft_model.embed_tokens.weight.requires_grad = False
> 
> # Load head
> head = torch.nn.Linear(model_args.hidden_size, model_args.vocab_size, bias=False)
> with open(os.path.join(path, "model.safetensors.index.json"), "r") as f:
>     index_json = json.loads(f.read())
>     head_path = index_json["weight_map"]["lm_head.weight"]
> with safe_open(os.path.join(path, head_path), framework="pt", device="cpu") as f:
>     tensor_slice = f.get_slice("lm_head.weight")
>     vocab_size, hidden_dim = tensor_slice.get_shape()
>     tensor = tensor_slice[:, :hidden_dim].float()
> 
> head.weight.data = tensor
> head.to("cuda:0")
> head.eval()
> 
> # -------------------------------- Load data --------------------------------
> 
> sharegpt_datapaths = list_local_files("/mnt/bn/kgvanasgcparnold/yinchangpan/datasets/eagle_data_qwen25_7b_full_data/sharegpt_grouped_5k/")
> ultra_chat_datapaths = list_local_files("/mnt/bn/kgvanasgcparnold/yinchangpan/datasets/eagle_data_qwen25_7b_full_data/ultrachat_0_199999_mufp16/")
> 
> combined_data_paths = (
>     sharegpt_datapaths[: int(len(sharegpt_datapaths) * 0.95)] + ultra_chat_datapaths
> )
> random.Random(42).shuffle(combined_data_paths)
> eval_data_paths = sharegpt_datapaths[int(len(sharegpt_datapaths) * 0.95) :][:100]
> 
> eagle_train_dataset = EagleLocalDataset(
>     combined_data_paths, transform=AddUniformNoise(std=0.5)
> )
> eagle_test_dataset = EagleLocalDataset(eval_data_paths)
> 
> eagle_collator = DataCollatorWithPadding()
> 
> # -------------------------------- Train --------------------------------
> 
> training_args = TrainingArguments(
>     output_dir=f"./hf_trainer_output_dir_qwen25_7b_fulldata_0604/{wandb_run_name}/",
>     num_train_epochs=10,
>     gradient_accumulation_steps=16,
>     per_device_train_batch_size=1,
>     per_device_eval_batch_size=1,
>     remove_unused_columns=False,
>     bf16=True,
>     fp16=False,
>     dataloader_num_workers=4,
>     warmup_ratio=0.01,
>     learning_rate=1e-4,  # 1e-3
>     lr_scheduler_type="constant",  # Placeholder, we override it in the trainer
>     max_grad_norm=0.5,  # 1
>     adam_beta1=0.9,  # 0.9
>     adam_beta2=0.95,  # 0.999
>     weight_decay=1e-2,
>     eval_strategy="steps",
>     logging_steps=32,
>     eval_steps=64,
>     save_strategy="steps",
>     save_steps=0.1,  # saves every 10% of training
>     save_total_limit=3,
> )
> 
> trainer = EagleTrainer(
>     model=draft_model,
>     head=head,
>     args=training_args,
>     train_dataset=eagle_train_dataset,
>     eval_dataset=eagle_test_dataset,
>     data_collator=eagle_collator,
>     min_lr_ratio=0.5,  # Custmer lr scheduler param
> )
> 
> trainer.train()

**Comment by KerwinKai** [comment]
@Greatpanc Can you print `vocab_size` and `hidden_dim`, meanwhile advice adjust the `intermediate_size` and others of draft model  config follow  origin model config (https://www.modelscope.cn/models/Qwen/Qwen2.5-7B-Instruct/file/view/master/config.json?status=1), we only need change `num_hidden_layers` to 1 for better training performance.
